whylog/config/rule.py

```diff
@@ -57,6 +65,265 @@ def get_causes_parsers(self):
     def get_effect_name(self):
         return self._effect.name
 
+    def get_search_ranges(self, effect_clues):
+        """
+        parser's search range is a dictionary with search range for its primary key type:
+        Sample parser's search range:
+        {
+            'date': {
+                InvesitgationStep.LEFT_BOUND: datetime(2016, 5, 29, 12, 33, 0),
+                InvesitgationStep.RIGHT_BOUND: datetime(2016, 5, 29, 12, 33, 30)
+            }
+        }
+        This method calculate search ranges for single rule based on its
+        constraints and effect clues.
+        Algorithm steps:
+            1. Check that effect parser has no empty primary key. If is empty
+             return NO_RANGE
+            2. Calculate search range for every cause parser.
+            3. Join parsers search ranges with this same log type
+        """
+        group_number, group_type = self._effect.get_primary_key_group()
+        if not group_number:
+            return self.NO_RANGE
+        parser_ranges = self._calculate_parsers_ranges(effect_clues, group_number, group_type)
+        if self._linkage == self.LINKAGE_OR:
+            self._update_parser_ranges_with_or_linkage(
+                effect_clues, group_number, group_type, parser_ranges
+            )
+        return self._aggregate_by_log_type(parser_ranges)
+
+    def _calculate_parsers_ranges(self, effect_clues, group, group_type):
+        """
+        This method calculate search range for every cause parser in rule.
+        This algorithm working only on delta constraints these allow reasoning about
+        parser's search range.
+        Delta constraints analysis:
+            Such constraint contains clues_groups list that contains two elements. Each of
+            elements is a two element list.
+            For example:
+                clues_groups = [[1,1], [0,1]]
+                It means that first group of first parser depend to first group o zero parser.
+                If these first groups of these parsers are in primary keys, we can predict
+                first parser search ranges basing on zero parser search range.
+                Assume that:
+                params = { min_delta: 10, max_delta: 100}
+                zero parser search ranges = {'int': {LEFT_BOUND: 200, RIGHT_BOUND: 250}
+                then
+                first parser search range = {'int': {LEFT_BOUND: 200 - 100, RIGHT_BOUND: 250 - 10}
+                In code first parser calls depended parser, zero parser calls base parser.
```

Comment:
Probably you meant: `is called` instead of `calls`, right? Be careful, it completely invert the meaning.